cmd/backup-manager/app/util/util.go

```diff
@@ -161,70 +161,103 @@ func GetOptionValueFromEnv(option, envPrefix string) string {
 	var args []string
-	config := backup.Spec.Dumpling
-	if config == nil {
+	config := backup.Spec
+
+	if config.TableFilter != nil && len(config.TableFilter) > 0 {
+		for _, tableFilter := range config.TableFilter {
+			args = append(args, "--filter", tableFilter)
+		}
+	} else {
+		if config.Dumpling != nil && config.Dumpling.TableFilter != nil && len(config.Dumpling.TableFilter) > 0 {
+			for _, tableFilter := range config.Dumpling.TableFilter {
+				args = append(args, "--filter", tableFilter)
+			}
+		} else {
+			args = append(args, defaultTableFilterOptions...)
+		}
+	}
+
+	if config.Dumpling == nil {
 		args = append(args, defaultOptions...)
-		args = append(args, defaultTableFilterOptions...)
 		return args
 	}
 
-	if len(config.Options) != 0 {
-		args = append(args, config.Options...)
+	if len(config.Dumpling.Options) != 0 {
+		args = append(args, config.Dumpling.Options...)
 	} else {
 		args = append(args, defaultOptions...)
 	}
 
-	if len(config.TableFilter) > 0 {
-		for _, tableFilter := range config.TableFilter {
-			args = append(args, "--filter", tableFilter)
-		}
-	} else {
-		args = append(args, defaultTableFilterOptions...)
-	}
 	return args
 }
 
 // ConstructBRGlobalOptionsForRestore constructs BR global options for restore.
 func ConstructBRGlobalOptionsForRestore(restore *v1alpha1.Restore) ([]string, error) {
 	var args []string
-	config := restore.Spec.BR
-	if config == nil {
+	config := restore.Spec
+	if config.BR == nil {
 		return nil, fmt.Errorf("no config for br in restore %s/%s", restore.Namespace, restore.Name)
 	}
-	args = append(args, constructBRGlobalOptions(config)...)
+	args = append(args, constructBRGlobalOptions(config.BR)...)
 	storageArgs, err := getRemoteStorage(restore.Spec.StorageProvider)
 	if err != nil {
 		return nil, err
 	}
 	args = append(args, storageArgs...)
-	if (restore.Spec.Type == v1alpha1.BackupTypeDB || restore.Spec.Type == v1alpha1.BackupTypeTable) && config.DB != "" {
-		args = append(args, fmt.Sprintf("--db=%s", config.DB))
-	}
-	if restore.Spec.Type == v1alpha1.BackupTypeTable && config.Table != "" {
-		args = append(args, fmt.Sprintf("--table=%s", config.Table))
+
+	if config.TableFilter != nil && len(config.TableFilter) > 0 {
+		for _, tableFilter := range config.TableFilter {
+			args = append(args, "--filter", tableFilter)
+		}
+	} else {
+		if ((restore.Spec.Type == v1alpha1.BackupTypeDB || restore.Spec.Type == v1alpha1.BackupTypeTable) && config.BR.DB != "") ||
+			restore.Spec.Type == v1alpha1.BackupTypeTable && config.BR.Table != "" {
+			if (restore.Spec.Type == v1alpha1.BackupTypeDB || restore.Spec.Type == v1alpha1.BackupTypeTable) && config.BR.DB != "" {
+				args = append(args, fmt.Sprintf("--db=%s", config.BR.DB))
+			}
+			if restore.Spec.Type == v1alpha1.BackupTypeTable && config.BR.Table != "" {
+				args = append(args, fmt.Sprintf("--table=%s", config.BR.Table))
+			}
+		} else {
+			args = append(args, defaultTableFilterOptions...)
```

Comment:
I think it's so complicated here. Can you simplify this part?

Reply:
You right this section is messy. I was trying to keep the previous functionality as much as possible but I will rework it.

Reply:
Trying to clean this up. How does BR work if database, tablename and filter are applied at the same time? Can I allow this to happen as that would make it a lot easier.

Reply:
See: https://github.com/pingcap/tidb-operator/pull/3059/files#r465578445

Reply:
No need to append the default filter option for BR.